### DISCUSSION TITLE
Question about workers graceful shutdown

### DISCUSSION THREAD

**Discussion by RobertoPrevato** [emmett-framework/granian#178] [Q&A]
Hi
I would like to listen to `SIGINT` and `SIGTERM` signals to stop loops when using long-polling or SSE, but these signals are not emitted when Granian stops workers.

For example, the following case works well with `uvicorn` and `hypercorn`, but not with `granian` because of that reason:
[https://github.com/Neoteroi/BlackSheep-Examples/tree/main/server-sent-events](https://github.com/Neoteroi/BlackSheep-Examples/tree/main/server-sent-events)

How can I know when Granian is about to stop workers to do some clean-up needed _before the ASGI lifespan stop event_ (use case discussed also [under this thread](https://github.com/encode/uvicorn/issues/1579#issuecomment-1419635974))?

**Comment by gi0baro** [comment]
Hello @RobertoPrevato,
this might be hard to assess at the moment, as the shutdown events are handled by Python but followed to the Rust side of things.

I have a question about the example you provided to better understand your use-case: what happens with Granian? Does it raise the `CancelledError` or it simply hangs forever and never exits?
Also, considering that `1.0` differs from previous versions in the shutdown handlers on the Rust side, it would also be nice if you try to compile the current master and check if something changes compared to 0.7.x.

**Reply by RobertoPrevato** [reply]
Hello @gi0baro
Thank You for replying so quickly! In my example, Granian does not raise any exception, it hangs forever. Note that uvicorn and Hypercorn behave in the same way, if one does not configure a signal handler. It is a bit of a egg and chicken situation, because Granian, uvicorn, and hypercorn are waiting for all connections to be closed or all pending tasks to complete, before communicating the application stop event (lifespan.shutdown in ASGI) but a web application using long-polling or server-sent events might need to do the same and close loops when the application is stopping. The trick mentioned in the uvicorn discussion I linked above is needed especially to take action _before_ the application stop event is fired by the ASGI server (in my opinion it would be best having 2 kinds of events for application stopping: before-stop and stop - or issue the  `lifespan.shutdown` message before awaiting for all pending tasks to be completed).
I will try to compile the current master of Granian and check how it behaves. PS. Congratulations for the project! I starred it a long time ago, it looks more and more interesting.

**Reply by RobertoPrevato** [reply]
I checked with the current master and `1.0` works much better because it does not hang and it closes on SIGINT.
The Python process does not get the SIGINT message, but when using the ASGI interface the app receives properly the `lifespan.shutdown` message. I would say everything looks good! :)

**Reply by gi0baro** [reply]
@RobertoPrevato thank you for testing 1.0!
I would say changing this in 0.x would be very hard, as the shutdown process in Hyper pre 1.0 has a few issues.

Regarding 1.0, probably we can experiment a bit with `signals` module and see if we can somehow re-emit the signal before the `lifespan.shutdown` message, I might have some time after closing the remaining issues in the 1.0 milestone. Of course if you have some time to play with it, I'd be happy to help you opening a PR or just supporting you in tests :)

**Reply by RobertoPrevato** [reply]
Thank You! I am willing to contribute but probably I won't have time soon. I don't know enough about Rush and I have little free time. But I will keep it in mind :)

**Reply by gi0baro** [reply]
@RobertoPrevato given the amount of changes, I released 1.0.0 without including this, which we can tackle properly for the next releases.
